lib/rolify/adapters/active_record/resource_adapter.rb

```diff
@@ -18,23 +18,26 @@ def resources_find(roles_table, relation, role_name)
           str
         end
 
+        id_column = id_column_name(relation)
+
         resources = relation.joins("INNER JOIN #{quote_table(roles_table)} ON #{quote_table(roles_table)}.resource_type IN (#{relations}) AND
-                                    (#{quote_table(roles_table)}.resource_id IS NULL OR #{quote_table(roles_table)}.resource_id = #{quote_table(relation.table_name)}.#{quote_column(relation.primary_key)})")
+                                    (#{quote_table(roles_table)}.#{quote_column(id_column)} IS NULL OR #{quote_table(roles_table)}.#{quote_column(id_column)} = #{quote_table(relation.table_name)}.#{quote_column(relation.primary_key)})")
         resources = resources.where("#{quote_table(roles_table)}.name IN (?) AND #{quote_table(roles_table)}.resource_type IN (?)", Array(role_name), klasses)
-        resources = resources.select("#{quote_table(relation.table_name)}.*")
         resources
       end
 
       def in(relation, user, role_names)
         roles = user.roles.where(:name => role_names).select("#{quote_table(role_class.table_name)}.#{quote_column(role_class.primary_key)}")
-        relation.where("#{quote_table(role_class.table_name)}.#{quote_column(role_class.primary_key)} IN (?) AND ((#{quote_table(role_class.table_name)}.resource_id = #{quote_table(relation.table_name)}.#{quote_column(relation.primary_key)}) OR (#{quote_table(role_class.table_name)}.resource_id IS NULL))", roles)
+        id_column = id_column_name(relation)
+
+        relation.where("#{quote_table(role_class.table_name)}.#{quote_column(role_class.primary_key)} IN (?) AND ((#{quote_table(role_class.table_name)}.#{quote_column(id_column)} = #{quote_table(relation.table_name)}.#{quote_column(relation.primary_key)}) OR (#{quote_table(role_class.table_name)}.#{quote_column(id_column)} IS NULL))", roles)
       end
 
       def applied_roles(relation, children)
         if children
-          relation.role_class.where('resource_type IN (?) AND resource_id IS NULL', self.relation_types_for(relation))
+          relation.role_class.where('resource_type IN (?) AND resource_id IS NULL AND resource_uuid IS NULL', self.relation_types_for(relation))
         else
-          relation.role_class.where('resource_type = ? AND resource_id IS NULL', relation.to_s)
+          relation.role_class.where('resource_type = ? AND resource_id IS NULL AND resource_uuid IS NULL', relation.to_s)
```

Comment:
I knot's a bigger change, but this also makes the generated sql a but safer since we avoid potential reference issue. Same would go for above as well.
```suggestion
          relation.role_class.where(resource_type: relation.to_s, resource_id: nil, resource_uuid: nil)
```